### ISSUE TITLE
Segmentation fault at attribute activity

### ISSUE THREAD

**Issue by aarontcopal2** [aarontcopal2/hpctoolkit#2]
I am getting a segmentation fault when running hpcrun ./myProgram
```
Program terminated with signal 11, Segmentation fault.
#0  0x00007f20c3206e96 in cstack_push (q=0x7f20bff2cf68, e=0x20ec8348e5894855) at ../../../../src/lib/prof-lean/stacks.c:223
223	    s_element_t *enext = (s_element_t *) atomic_load(&e->Ad(next));
```

Stacktrace
```
#0  0x00007f20c3206e96 in cstack_push (q=0x7f20bff2cf68, e=0x20ec8348e5894855) at ../../../../src/lib/prof-lean/stacks.c:223
#1  0x00007f20c3207015 in bistack_push (s=0x7f20bff2cf68, e=0x7f20b05d97d8) at ../../../../src/lib/prof-lean/bistack.c:85
#2  0x00007f20c32070f5 in bichannel_push (ch=0x7f20bff2cf68, dir=bichannel_direction_forward, e=0x7f20b05d97d8)
    at ../../../../src/lib/prof-lean/bichannel.c:81
#3  0x00007f20c31dbc73 in gpu_activity_t_bichannel_push (c=0x7f20bff2cf68, dir=bichannel_direction_forward, e=0x7f20b05d97d8)
    at ../../../../src/tool/hpcrun/gpu/gpu-activity-channel.c:108
#4  0x00007f20c31dbe10 in gpu_activity_channel_produce (channel=0x7f20bff2cf68, a=0x7f20b1ffcf10)
    at ../../../../src/tool/hpcrun/gpu/gpu-activity-channel.c:157
#5  0x00007f20c31dbf75 in attribute_activity (hc=0x7f20b05d9938, activity=0x7f20b1ffcf10, cct_node=0x7f20bff2cf88)
    at ../../../../src/tool/hpcrun/gpu/gpu-activity-process.c:139
#6  0x00007f20c31dc07a in gpu_memcpy_process (activity=0x7f20b1ffcf10) at ../../../../src/tool/hpcrun/gpu/gpu-activity-process.c:186
#7  0x00007f20c31dc814 in gpu_activity_process (ga=0x7f20b1ffcf10) at ../../../../src/tool/hpcrun/gpu/gpu-activity-process.c:585
#8  0x00007f20c31f4edc in opencl_activity_process (event=0x294f650, user_data=0x293ae90) at ../../../../src/tool/hpcrun/gpu/intel/opencl-api.c:77
#9  0x00007f20c31f4e9a in opencl_buffer_completion_callback (event=0x294f650, event_command_exec_status=0, user_data=0x293ae90)
    at ../../../../src/tool/hpcrun/gpu/intel/opencl-api.c:65
#10 0x00007f20bdd37e4c in ?? () from /lib64/libnvidia-opencl.so.1
#11 0x00007f20bdd3edae in ?? () from /lib64/libnvidia-opencl.so.1
#12 0x00007f20bdd3f650 in ?? () from /lib64/libnvidia-opencl.so.1
#13 0x00007f20bde69b18 in ?? () from /lib64/libnvidia-opencl.so.1
#14 0x00007f20c2f8d35b in monitor_thread_fence2 () at pthread.c:978
#15 0x00007f20c1d6eea5 in start_thread () from /lib64/libpthread.so.0
#16 0x00007f20c22858cd in clone () from /lib64/libc.so.6
```

These are the locations of calls to activity and correlation channels.

https://github.com/aarontcopal2/hpctoolkit/blob/ab33ace54e8430ea79af5d097c39987caad4e6a8/src/tool/hpcrun/gpu/intel/opencl-api.c#L56

https://github.com/aarontcopal2/hpctoolkit/blob/ab33ace54e8430ea79af5d097c39987caad4e6a8/src/tool/hpcrun/gpu/intel/opencl-api.c#L58

(internally calls gpu_correlation_channel_produce)
https://github.com/aarontcopal2/hpctoolkit/blob/ab33ace54e8430ea79af5d097c39987caad4e6a8/src/tool/hpcrun/gpu/intel/opencl-api.c#L70

https://github.com/aarontcopal2/hpctoolkit/blob/8b92e780f95dae42df56834aa6e3693c00d14739/src/tool/hpcrun/gpu/gpu-activity-process.c#L139

Are the function calls and arguments here correct? I feel the segfault is related to one of these calls being incorrect

**Comment by Jokeren** [comment]
https://github.com/aarontcopal2/hpctoolkit/blob/ab33ace54e8430ea79af5d097c39987caad4e6a8/src/tool/hpcrun/gpu/intel/opencl-api.c#L32

It shouldn't be called in the application thread. 

https://github.com/aarontcopal2/hpctoolkit/blob/ab33ace54e8430ea79af5d097c39987caad4e6a8/src/tool/hpcrun/gpu/gpu-activity-process.c#L151

The lookup operation and the insert operation should both be called by the background thread only.

https://github.com/aarontcopal2/hpctoolkit/blob/ab33ace54e8430ea79af5d097c39987caad4e6a8/src/tool/hpcrun/gpu/intel/opencl-api.c#L31

Fix the correlation id problem.

And `intel.c` is still missing.

https://github.com/aarontcopal2/hpctoolkit/blob/ab33ace54e8430ea79af5d097c39987caad4e6a8/src/tool/hpcrun/gpu/intel/opencl-intercept.c#L40

The callback should called before the actual function.

**Comment by aarontcopal2** [comment]
1. Now the lookup, insert and produce/consume function calls are made by the background threads
2. I have made the correlation_id fetching and increments atomic.
3. intel.c has not been added yet. I will have to use device_notify_map() inside device_initializers.c because I need to dlopen libraries and check if certain functions are present. So I dont see the point of adding a "intel.c" file in sample-sources
4. In Intel's opencl, you can register callbacks for an event only after that event has been submitted. So, I cant change the order of lines here. Here is a [link](https://stackoverflow.com/questions/54331860/opencl-callback-on-already-completed-event) with more details

Sadly, the segmentation faults keep occuring event after making these changes.

**Comment by Jokeren** [comment]
@aarontcopal2 , `opencl_subscriber_callback` is not a callback for a event. It's just your own function, you can put it anywhere you like. But if you are sure performance events won't be returned before `clSetEventCallback`, it is fine to leave the `opencl_subscriber_callback` unchanged.

**Comment by aarontcopal2** [comment]
I have made opencl_subscriber_callback as a callback because some of the calls made by it (gpu_correlation_id_map_insert, gpu_activity_channel_consume, gpu_correlation_channel_produce) needed to be made from background threads, right?

**Comment by Jokeren** [comment]
No, `opencl_subscriber_callback` is called by the application thread. It is a callback for the function. In contrast, `clSetEventCallback` sets the (event) callback for buffer completion, which should be executed by the background thread. No sure if you previously had a different understanding.

**Comment by Jokeren** [comment]
Just noticed two suspicious lines.

https://github.com/aarontcopal2/hpctoolkit/blob/ab33ace54e8430ea79af5d097c39987caad4e6a8/src/tool/hpcrun/gpu/intel/opencl-intercept.c#L38

You have to fix this later after you fix the current bug.

Frequent *malloc* winds up wasting time and memory. The correct way to do it is reusing a piece of memory. You can reach out to John to discuss the design of this feature.

https://github.com/HPCToolkit/hpctoolkit/blob/master/src/tool/hpcrun/gpu/nvidia/cupti-activity-translate.c#L652

A `cstack_ptr_set` is missing for gpu activities in your case. I suppose this is the main reason for your crash.

**Comment by aarontcopal2** [comment]
https://github.com/aarontcopal2/hpctoolkit/blob/280e8e004658e48d7d934068805ccbef87d48524/src/tool/hpcrun/gpu/intel/opencl-intercept.c#L46

Even opencl_subscriber_callback is being called by clSetEventCallback

**Comment by Jokeren** [comment]
@aarontcopal2 , they are different. Notice the event name is called `CL_SUBMITTED`. The background thread does not submit any kernels.

**Comment by aarontcopal2** [comment]
@Jokeren ok, that is a possibility. If thats the case, gpu_correlation_id_map_insert, gpu_activity_channel_consume and gpu_correlation_channel_produce are being called by the main thread. Is this ok or should we try moving these calls to a background thread

**Comment by Jokeren** [comment]
`gpu_correlation_id_map_insert` should be called by the background thread.
`gpu_correlation_channel_produce` should be called by the application thread.
`gpu_activity_channel_consume` should be called by the application thread.

**Comment by aarontcopal2** [comment]
@Jokeren your guess was spot-on. Not calling cstack_ptr_set() was the reason of the segfaults.

@jmellorcrummey , should'nt call to cstack_ptr_set() be abstracted? It seems like a low-level detail that should'nt be called between activity_translate() and activity_process() calls

**Comment by aarontcopal2** [comment]
Let me create another issue for keeping track of missing gpu_metrics. Ill close this issue of segfaults once John replies about cstack_ptr_set

**Comment by riyaj0207** [comment]
hello,
i am gettin this error in map "Segmentation fault.  ?? () from /lib64/libstdc++.so.6"

this is my code 


  template<typename T>
  struct Allocator : allocator<T> {
      inline typename allocator<T>::pointer allocate(typename allocator<T>::size_type n, typename allocator<int>::const_pointer = 0) {

      return reinterpret_cast<typename allocator<T>::pointer>(malloc (n * sizeof (T)));
      }

      inline void deallocate(typename allocator<T>::pointer p, typename allocator<T>::size_type n) {

          free(p);
      }

      template<typename U>
      struct rebind {
          typedef Allocator<U> other;
      };

   Allocator() =default;

    template<class Other>
    Allocator( const Allocator<Other>& _Right ) {}
  ~Allocator()=default;
  };
static std::map<void*,int,std::less<void*>,Allocator<std::pair<const void*,int> > > sizemap;

 void map_insert(void* p,int size)

 {

static sizemap.insert(std::pair<void*,int>(p,size)); }


int  map_delete(void* p)
{
return sizemap[p];

}

void* operator new (size_t size)
    {std::cout<<"overloaded new  "<<size<<endl;
  memused += size;
  cout<<memused<<"memused"<<endl;
    void *p = malloc(size);
  map_insert(p,size);
    return p;}

    void operator delete(void *p)
    {std::cout<<"delete"<<endl;

  memused-=map_delete(p);
  cout<<memused<<"  memused "<<endl;
   free(p);

  }
can anyone help me, please?

**Comment by aarontcopal2** [comment]
@riyaj0207 If you have issues with hpctoolkit, could you raise a separate issue at https://github.com/HPCToolkit/hpctoolkit/issues? This is a repository that I have cloned, not the right place to raise issues

**Comment by riyaj0207** [comment]
m sorry.. m new to GitHub..ill delete everything.. as nothing happened :D
